sdks/java/fn-execution/src/main/java/org/apache/beam/sdk/fn/data/BeamFnDataOutboundAggregator.java

```diff
@@ -0,0 +1,232 @@
+  private final int sizeLimit;
+  private final long timeLimit;
+  private final Map<LogicalEndpoint, Coder<?>> outputLocations;
+  private final StreamObserver<Elements> outboundObserver;
+  private final Map<LogicalEndpoint, ByteString.Output> buffer;
+  @VisibleForTesting ScheduledFuture<?> flushFuture;
+  private long byteCounter;
+
+  public BeamFnDataOutboundAggregator(
+      PipelineOptions options, StreamObserver<BeamFnApi.Elements> outboundObserver) {
+    this.sizeLimit = getSizeLimit(options);
+    this.timeLimit = getTimeLimit(options);
+    this.outputLocations = new HashMap<>();
+    this.buffer = new ConcurrentHashMap<>();
+    this.outboundObserver = outboundObserver;
+    if (timeLimit > 0) {
+      this.flushFuture =
+          Executors.newSingleThreadScheduledExecutor(
+                  new ThreadFactoryBuilder()
+                      .setDaemon(true)
+                      .setNameFormat("DataBufferOutboundFlusher-thread")
+                      .build())
+              .scheduleAtFixedRate(
+                  this::periodicFlush, timeLimit, timeLimit, TimeUnit.MILLISECONDS);
+    } else {
+      this.flushFuture = null;
+    }
+  }
+
+  /**
+   * Register the outbound logical endpoint alongside it's value coder. All registered endpoints
+   * will eventually be removed when {@link #close()} is called at the end of a bundle.
+   */
+  public <T> void registerOutputLocation(LogicalEndpoint endpoint, Coder<T> coder) {
+    LOG.debug("Registering endpoint: {}", endpoint);
+    outputLocations.put(endpoint, coder);
+  }
+
+  public void flush() throws IOException {
+    if (byteCounter > 0) {
+      outboundObserver.onNext(convertBufferForTransmission().build());
+    }
+  }
+
+  /**
+   * Closes the streams for all registered outbound endpoints. Should be called at the end of each
+   * bundle.
+   */
+  @Override
+  public void close() {
+    LOG.debug("Closing streams for outbound endpoints {}", outputLocations);
+    if (byteCounter == 0 && outputLocations.isEmpty()) {
+      return;
+    }
+    Elements.Builder bufferedElements = convertBufferForTransmission();
+    for (LogicalEndpoint outputLocation : outputLocations.keySet()) {
+      if (outputLocation.isTimer()) {
+        bufferedElements
+            .addTimersBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setTimerFamilyId(outputLocation.getTimerFamilyId())
+            .setIsLast(true);
+      } else {
+        bufferedElements
+            .addDataBuilder()
+            .setInstructionId(outputLocation.getInstructionId())
+            .setTransformId(outputLocation.getTransformId())
+            .setIsLast(true);
+      }
+    }
+    outboundObserver.onNext(bufferedElements.build());
+    outputLocations.clear();
+  }
+
+  public <T> void accept(LogicalEndpoint endpoint, T data) throws Exception {
```

Comment:
In a follow-up PR it will make sense to make give a `ThrowingConsumer` to the BeamFnDataWriteRunner so we can avoid the map lookup on each element call. If its trivial, feel free to do now.

Reply:
Done.